pallet-chainlink-feed/src/lib.rs

```diff
@@ -0,0 +1,1345 @@
+	type Event: From<Event<Self>> + Into<<Self as frame_system::Trait>::Event>;
+
+	/// Type for feed indexing.
+	type FeedId: Member + Parameter + Default + Copy + HasCompact + BaseArithmetic;
+
+	/// Type for round indexing.
+	type RoundId: Member + Parameter + Default + Copy + HasCompact + BaseArithmetic + Into<u64>;
+
+	/// Oracle feed values.
+	type Value: Member + Parameter + Default + Copy + HasCompact + PartialEq + BaseArithmetic;
+
+	/// Interface used for balance transfers.
+	type Currency: ReservableCurrency<Self::AccountId>;
+
+	/// The module id used to determine the account for storing the funds used to pay the oracles.
+	type ModuleId: Get<ModuleId>;
+
+	/// The minimum amount of funds that need to be present in the fund account.
+	type MinimumReserve: Get<BalanceOf<Self>>;
+
+	/// Maximum allowed string length.
+	type StringLimit: Get<u32>;
+
+	/// Maximum number of oracles per feed.
+	type OracleCountLimit: Get<u32>;
+
+	/// Maximum number of feeds.
+	type FeedLimit: Get<Self::FeedId>;
+
+	/// Number of rounds to keep around per feed.
+	type PruningWindow: Get<Self::RoundId>;
+
+	// type WeightInfo: WeightInfo;
+}
+
+/// The configuration for an oracle feed.
+#[derive(Clone, Encode, Decode, Default, Eq, PartialEq, RuntimeDebug)]
+pub struct FeedConfig<
+	AccountId: Parameter,
+	Balance: Parameter,
+	BlockNumber: Parameter,
+	RoundId: Parameter,
+	Value: Parameter,
+> {
+	owner: AccountId,
+	pending_owner: Option<AccountId>,
+	submission_value_bounds: (Value, Value),
+	submission_count_bounds: (u32, u32),
+	payment: Balance,
+	timeout: BlockNumber,
+	decimals: u8,
+	description: Vec<u8>,
+	restart_delay: RoundId,
+	reporting_round: RoundId,
+	latest_round: RoundId,
+	first_valid_round: Option<RoundId>,
+	oracle_count: u32,
+}
+pub type FeedConfigOf<T> = FeedConfig<
+	<T as frame_system::Trait>::AccountId,
+	BalanceOf<T>,
+	<T as frame_system::Trait>::BlockNumber,
+	<T as Trait>::RoundId,
+	<T as Trait>::Value,
+>;
+
+/// Round data relevant to consumers.
+/// Will only be constructed once minimum amount of submissions have
+/// been provided.
+#[derive(Clone, Encode, Decode, Default, Eq, PartialEq, RuntimeDebug)]
+pub struct Round<BlockNumber: Parameter, RoundId: Parameter, Value: Parameter> {
+	started_at: BlockNumber,
+	answer: Option<Value>,
+	updated_at: Option<BlockNumber>,
+	answered_in_round: Option<RoundId>,
+}
+pub type RoundOf<T> = Round<<T as frame_system::Trait>::BlockNumber, <T as Trait>::RoundId, <T as Trait>::Value>;
+
+impl<B, R, V> Round<B, R, V>
+where
+	B: Parameter + Default, // BlockNumber
+	R: Parameter + Default, // RoundId
+	V: Parameter + Default, // Value
+{
+	/// Create a new Round with the given starting block.
+	fn new(started_at: B) -> Self {
+		Self {
+			started_at,
+			..Default::default()
+		}
+	}
+}
+
+/// Round data relevant to oracles.
+#[derive(Clone, Encode, Decode, Default, Eq, PartialEq, RuntimeDebug)]
+pub struct RoundDetails<Balance: Parameter, BlockNumber: Parameter, Value: Parameter> {
```

Comment:
```suggestion
pub struct RoundDetails<Balance, BlockNumber, Value> {
```

why bound them here? Just saying bc it is not common in substrate.

Reply:
I thought they were necessary somehow :grimacing: